### PR TITLE
GWI Challenge, by Archontellis Rafail Sotirchellis

### PR DESCRIPTION
**Pull request by arsotirchellis**

### CODE REVIEW

**Comment by naimis20** [comment]
Good morning mate! I am going to make some comments in your PR , please respond in your convenience!

File: src/controllers/middleware.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	"github.com/dgrijalva/jwt-go/request"
+	"github.com/gin-gonic/gin"
+)
+
+func removeBearer(token string) (string, error) {
+	if len(token) > 6 && strings.ToUpper(token[0:7]) == "BEARER " {
+		return token[7:], nil
+	}
+	return token, nil
+}
```
Comment (naimis20) [inline src/controllers/middleware.go]: Why do you remove Bearer?

File: src/data/models/audience.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	responseLiteral += audienceToCompose.Gender.String(true)
+	responseLiteral += " from "
+	responseLiteral += strconv.FormatUint(uint64(audienceToCompose.AgeGroupLowerLimit), 10) + "-" + strconv.FormatUint(uint64(audienceToCompose.AgeGroupUpperLimit), 10)
+
+	if audienceToCompose.AudienceInfo.AudienceInfoType == HoursOnSocialMedia {
+		responseLiteral += " that spent more than " + strconv.FormatUint(uint64(audienceToCompose.AudienceInfo.AudienceInfoStat), 10) + " hours on social media daily"
+	} else {
+		responseLiteral += " that make more than " + strconv.FormatUint(uint64(audienceToCompose.AudienceInfo.AudienceInfoStat), 10) + " purchases per month"
+	}
+	return responseLiteral
```
Comment (naimis20) [inline src/data/models/audience.go]: It could be cleaner if you have used the standard formatter package here called "fmt"

File: src/serializers/audiences_serializers.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+	*models.Audience
+}
+
+type AudienceResponse struct {
+	Title           string `json:"title"`
+	AudienceLiteral string `json:"audience"`
+}
+
+func NewAudienceSerializer() AudienceSerializer {
+	return AudienceSerializer{}
```
Comment (naimis20) [inline src/serializers/audiences_serializers.go]: In general you should either choose new(AudienceSerializer) vs AudienceSerializer{} to be consistent throughout the project.

**Comment by naimis20** [comment]
Another thing is that it would be better to set the userID as guid instead of uint.

**Comment by naimis20** [comment]
Moreover in general, all your entities should have an ID property and its type should be guid.

**Comment by naimis20** [comment]
On Describe Chart call , in case case a chart is not favorite, the following is the result:
`{
    "errors": {
        "db_error": "User has not favorited this chart"
    }
}`
however this exposes the nature of the problem which can inform a potential attacker of how the system works.

File: src/services/users_service_test.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
+package services
+
+import (
+	"gwi-challenge/data/models"
+	"testing"
+)
+
+func TestEncryption(t *testing.T) {
```
Comment (teliaz) [inline src/services/users_service_test.go]: I see that there are some passing tests in your PR. 
In a large scale project how would you approach the testing implementation and workflow?
What are in your opinion the benefits of testing?

**Comment by arsotirchellis** [comment]
> Good morning mate! I am going to make some comments in your PR , please respond in your convenience!

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+COPY src/ /app
+WORKDIR /app
+
+RUN go build
+
+CMD ["./gwi-challenge"]
+
+# docker build -t gwi-demo .
+# docker run -p 8080:8080 gwi-demo
```
[the remark this answers is not in the document]
Comment (arsotirchellis) [reply Dockerfile]: No, they have been forgotten there. There were useful during Dockerfile creation.

File: src/common/config.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+	config := Configuration{}
+	err := gonfig.GetConf("./config.json", &config)
+	if err != nil {
+		panic(err)
+	}
+	configuration = &config
+}
+
+func GetConfig() *Configuration {
+	return configuration
```
[the remark this answers is not in the document]
Comment (arsotirchellis) [reply src/common/config.go]: Good point. I prefer your approach because with the current implementation we can initialize the configuration multiple times.

File: src/config.json [hunk 7]
```diff
@@ -0,0 +1,8 @@
+{
+    "Secret": "jhsbd@%&^G4G$%6y^&*Jdfada3r3f5dfdf##%^",
+    "IsDemo": true,
+    "DbPath": "./gwi_demo.db",
+    "SqlDropPath": "./data/sql/db_drop.sql",
+    "SqlCreatePath": "./data/sql/db_create.sql",
+    "SqlPopulatePath": "./data/sql/db_populate.sql"
+}
```
[the remark this answers is not in the document]
Comment (arsotirchellis) [reply src/config.json]: It’s not safe or protected. Also` "Secret"  `property is not encrypted for sure. One good way to go is to use [Docker secrets](https://docs.docker.com/engine/swarm/secrets/). I have not prior experience with these to speak in more details, but it seems that it solves my problem.

File: src/controllers/audiences_routers.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+	userID := context.GetInt("user_id")
+	if err := services.UnfavoriteAudience(uint(userID), title); err != nil {
+		context.JSON(http.StatusBadRequest, common.NewError("db_error", err))
+		return
+	}
+}
+
+func describeAudience(context *gin.Context) {
+	title := context.Param("title")
+	userID := context.GetInt("user_id")
```
[the remark this answers is not in the document]
Comment (arsotirchellis) [reply src/controllers/audiences_routers.go]: `userID` is stored in the request’s context by me (middleware.go, line 45) so it will always be valid. Title should have a max allowed length. So, I agree that title needs a validator.

File: src/controllers/charts_routers.go [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+)
+
+func ChartsRoutesRegister(router *gin.RouterGroup) {
+	router.POST("", createChart)
+	router.GET("/:title", getChart)
+	router.PUT("/:title", updateChart)
+	router.DELETE("/:title", deleteChart)
+	router.POST("/:title/favorite", favoriteChart)
+	router.DELETE("/:title/favorite", unfavoriteChart)
+	router.PATCH("/:title/describe", describeChart)
```
[the remark this answers is not in the document]
Comment (arsotirchellis) [reply src/controllers/charts_routers.go]: My first thought was that I must not expose the table’s primary key for security reasons. Title, in retrospect, was a bad decision. It has spaces, in some cases has too many letters etc. I have made a todo note about this topic in Readme file. I believe that the best practice is to provide the identifier to the client encrypted and decrypt it in every request.

File: src/controllers/charts_routers.go [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+func unfavoriteChart(context *gin.Context) {
+	title := context.Param("title")
+	userID := context.GetInt("user_id")
+	if err := services.UnfavoriteChart(uint(userID), title); err != nil {
+		context.JSON(http.StatusBadRequest, common.NewError("db_error", err))
+		return
+	}
+}
```
[the remark this answers is not in the document]
Comment (arsotirchellis) [reply src/controllers/charts_routers.go]: Yes. The code could be the half this way.

File: src/controllers/middleware.go (hunk 1 above)
Comment (arsotirchellis) [reply src/controllers/middleware.go]: Because the format of the authorization header is like: 
`BEARER eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHBpcnkiOjE1NzYzNDc3OTAsImlkIjox`
The JWT token, which I need, is the second part of the above literal.

File: src/data/models/audience.go (hunk 2 above)
Comment (arsotirchellis) [reply src/data/models/audience.go]: Its true.

File: src/serializers/audiences_serializers.go (hunk 3 above)
Comment (arsotirchellis) [reply src/serializers/audiences_serializers.go]: I think that it has the same result. But I agree with you. In large scale projects consistency in code style is a must.

File: src/services/users_service_test.go (hunk 4 above)
Comment (arsotirchellis) [reply src/services/users_service_test.go]: Testing, in large scale projects is a must. It helps you to find problems early, to make your code better (use interfaces, split your methods etc.) and handle weird situations which you have never thought about (null variables, edge values etc.). With testing you can prevent someone to change in the wrong way the code. Unit tests also are a very good documentation for a future reader. 
In my opinion a developer must write unit tests in parallel with the source code (I have not experience in TDD to say if its better or not) because they are helping in debugging too.
Pull requests on the main git branch should have a rule for passing tests and a rule for a minimum coverage percentage for touched and new files.

**Comment by arsotirchellis** [comment]
> Moreover in general, all your entities should have an ID property and its type should be guid.

I agree with you. It should be a Bad Request response code without the “db_error” in response literal.

**Comment by arsotirchellis** [comment]
> On Describe Chart call (as well as other calls too) , in case case a chart is not favorite, the following is the result:
> `{ "errors": { "db_error": "User has not favorited this chart" } }`
> however this exposes the nature of the problem which can inform a potential attacker of how the system works.

I agree with you. It should be a Bad Request response code without the “db_error” in response literal.

**Comment by arsotirchellis** [comment]
> Moreover in general, all your entities should have an ID property and its type should be guid.

UUID’s have disadvantages in performance, ordering, storage and debugging. On the other hand, they are more secure, less predictable and they have a value and not a pseudo-value. The last means that they are better for a distributed system with multiple databases. If you implement for example the CQRS pattern you have two views or tables for the same entity. So, I agree with you. I implemented it with Auto Increment Keys in favor of simplicity.

**Comment by teliaz** [comment]
Thank you for taking the time to implement this challenge @arsotirchellis.
